pageserver/src/walreceiver.rs

```diff
@@ -468,14 +468,17 @@ async fn timeline_wal_broker_loop_step(
             // finally, if no other tasks are completed, get another broker update and possibly reconnect
             updates = broker_subscription.fetch_data() => match updates {
                 Some(mut all_timeline_updates) => {
-                    if let Some(subscribed_timeline_updates) = all_timeline_updates.remove(&id) {
-                        match wal_connection_manager.select_connection_candidate(subscribed_timeline_updates) {
-                            Some(candidate) => {
-                                info!("Switching to different safekeeper {} for timeline {id}, reason: {:?}", candidate.safekeeper_id, candidate.reason);
-                                wal_connection_manager.change_connection(candidate.safekeeper_id, candidate.wal_producer_connstr).await;
-                            },
-                            None => {}
+                    match all_timeline_updates.remove(&id) {
+                        Some(subscribed_timeline_updates) => {
+                            match wal_connection_manager.select_connection_candidate(subscribed_timeline_updates) {
+                                Some(candidate) => {
+                                    info!("Switching to different safekeeper {} for timeline {id}, reason: {:?}", candidate.safekeeper_id, candidate.reason);
+                                    wal_connection_manager.change_connection(candidate.safekeeper_id, candidate.wal_producer_connstr).await;
+                                },
+                                None => debug!("No connection candidate was selected for timeline"),
+                            }
                         }
+                        None => warn!("Timeline has an active broker subscription, but got no updates. Other data length: {}", all_timeline_updates.len()),
```

Comment:
How this is possible? If it doesn't, let's add comment/assert about being unreachable.

Reply:
I've just looked at the logs, and looks like we've got a conflict of a kind, that triggers exactly this: the only failed test currently, https://app.circleci.com/pipelines/github/neondatabase/neon/6914/workflows/46edb373-663a-4863-9d09-f460e36e68fc/jobs/70757, sometimes receives `WAL_BACKUP` keys only, since those share the same key prefix, as the one we subscribe to:

```
2022-06-02T11:21:51.798462Z ERROR wal_receiver_main:timeline{id=ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80}:etcd_subscription:etcd_broker: Failed to parse timeline update: Error during parsing etcd data: KV has unexpected key 'neon/ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80/safekeeper/WAL_BACKUP/694d81241b4118c7' that does not match required regex ^neon/ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80/safekeeper/([[:digit:]]+)$
2022-06-02T11:21:51.798737Z ERROR wal_receiver_main:timeline{id=ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80}:etcd_subscription:etcd_broker: Failed to parse timeline update: Error during parsing etcd data: KV has unexpected key 'neon/ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80/safekeeper/WAL_BACKUP/694d81241b4118c5' that does not match required regex ^neon/ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80/safekeeper/([[:digit:]]+)$
2022-06-02T11:21:51.798914Z  WARN wal_receiver_main:timeline{id=ddf794d7b424414ba62dbab6afe69829/49ec01927debff5d347a6aaa96420a80}: Timeline has an active broker subscription, but got no updates. Other data length: 0
```

should we split those keys entirely, to avoid overlap in the same subscription?

Reply:
Yes, let's skip `safekeeper`, adjusting `get_campaign_name`.